### PR TITLE
8303027: [Lilliput/JDK17] Correctly resolve forwarded objects in G1 heap iteration

### PR DESCRIPTION
**Pull request by rkennke**
A user provided a (pretty simple) test case that makes the VM crash with Lilliput, when it should exit (somewhat more) gracefully with an OOME. The reason for the crash is that in G1, object_iterate() (or rather, the block_size() method that it calls) does not correctly resolve possibly forwarded objects, and then crashes because it accesses the Klass*, which is overridden by the forwarding pointer.

The trouble is that, we don't need (or even can't) resolve forwarded objects during full-GC, because there we are using sliding-forwarding which is designed precisely to preserve the object's Klass* while being forwarded. Outside of full-GC we don't need to preserve the Klass* because the Klass* is preserved in the forwarded copy.

This bug only exists in Lilliput/JDK17, the code path that leads to object_iterate() (restoring self-forwarded objects at evac failure) doesn't exist anymore in later versions - it uses a better way to iterate over the relevant self-forwarded objects.

The proposed fix is to do two different loops, one that resolves the forwardees when needed, and the other one that doesn't. The alternative would have to check for the condition in_full_gc() on every object being scanned, which would affect performance. I included the test case as jtreg test.

Testing:
 - [x] runtime/oom/TestOOM.java
 - [x] tier1
 - [ ] tier2

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303027](https://bugs.openjdk.org/browse/JDK-8303027): [Lilliput/JDK17] Correctly resolve forwarded objects in G1 heap iteration

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [590742c5](https://git.openjdk.org/lilliput-jdk17u/pull/6/files/590742c5bb0e979e191fa9cb320609de8b237916)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/lilliput-jdk17u pull/6/head:pull/6` \
`$ git checkout pull/6`

Update a local copy of the PR: \
`$ git checkout pull/6` \
`$ git pull https://git.openjdk.org/lilliput-jdk17u pull/6/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6`

View PR using the GUI difftool: \
`$ git pr show -t 6`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/lilliput-jdk17u/pull/6.diff">https://git.openjdk.org/lilliput-jdk17u/pull/6.diff</a>

</details>

### CODE REVIEW

**Review by shipilev — APPROVED** [review]
Looks okay. 

But depending on the frequency of `block_is_obj(p)` being true, we might just put the `in_full_gc()` condition inside the `block_size()` itself?

**Comment by rkennke** [comment]
> But depending on the frequency of `block_is_obj(p)` being true, we might just put the `in_full_gc()` condition inside the `block_size()` itself?

The actual problem only arises when evacuation fails and GC tries to recover from that (specifically, when the GC is restoring self-forwarded object headers). That is not very performance critical, I think. However, block_size() is called from many other places during GC, and is potentially applied to all (live) objects, e.g. during heap scan. It may be helped a little by bitmap based iteration, but I'd rather not risk affecting performance of all the common paths, if I can avoid it.

**Comment by rkennke** [comment]
> Looks okay.
> 
> But depending on the frequency of `block_is_obj(p)` being true, we might just put the `in_full_gc()` condition inside the `block_size()` itself?

I just pushed an attempt to use a templated loop and block_size(), that avoids the duplication AND should perform well. WDYT?

File: src/hotspot/share/gc/g1/heapRegion.hpp [hunk 1]
```diff
@@ -183,6 +186,8 @@ class HeapRegion : public CHeapObj<mtGC> {
 
   // Returns the object size for all valid block starts
   // and the amount of unallocated words if called on top()
+  template<bool RESOLVE>
+  size_t block_size(const HeapWord* p) const;
```
Comment (shipilev) [inline src/hotspot/share/gc/g1/heapRegion.hpp:190]: There is non-templated `block_size` here too, just below. Is the non-templated version used? If so, do you want to drop it and say `template<bool RESOLVE = true>` instead?

**Review by shipilev — APPROVED** [review]
This looks much better, thanks.

**Review by shipilev — APPROVED** [review]

**Pull request closed without merging**